nipst/nipst.go

```diff
@@ -0,0 +1,380 @@
+}
+
+func (n *NIPST) load() {
+	// TODO: implement
+}
+
+func (n *NIPST) Valid() bool {
+	// TODO: implement
+	return true
+}
+
+type ActivationBuilder interface {
+	BuildActivationTx(proof NIPST)
+}
+
+type NIPSTBuilder struct {
+	id                []byte
+	space             Space
+	duration          SeqWorkTicks
+	postProver        PostProver
+	poetProver        PoetProvingService
+	activationBuilder ActivationBuilder
+	stop              bool
+	stopM             sync.Mutex
+	errChan           chan error
+
+	nipst *NIPST
+}
+
+func NewNIPSTBuilder(id []byte, space Space, duration SeqWorkTicks, postProver PostProver,
+	poetProver PoetProvingService, activationBuilder ActivationBuilder) *NIPSTBuilder {
+	return &NIPSTBuilder{
+		id:                id,
+		space:             space,
+		duration:          duration,
+		postProver:        postProver,
+		poetProver:        poetProver,
+		activationBuilder: activationBuilder,
+		stop:              false,
+		errChan:           make(chan error),
+		nipst:             initialNIPST(id, space, duration, nil),
+	}
+}
+
+func (nb *NIPSTBuilder) stopped() bool {
+	nb.stopM.Lock()
+	defer nb.stopM.Unlock()
+	return nb.stop
+}
+
+func (nb *NIPSTBuilder) Stop() {
+	nb.stopM.Lock()
+	nb.stop = true
+	defer nb.stopM.Unlock()
+}
+
+func (nb *NIPSTBuilder) Start() {
+	nb.stopM.Lock()
+	nb.stop = false
+	go nb.loop()
+	defer nb.stopM.Unlock()
+}
+
+func (nb *NIPSTBuilder) loop() {
+	defTimeout := 1 * time.Second // temporary solution
+	nb.nipst.load()
+
+	// Phase 0
+	if nb.nipst.postCommitment == nil {
+		log.Info("starting PoST initialization, id: %x, space: %v",
+			nb.id, nb.space)
+
+		com, err := nb.postProver.initialize(nb.id, nb.space, defTimeout)
+		if err != nil {
+			nb.errChan <- fmt.Errorf("failed to initialize PoST: %v", err)
+			return
+		}
+		if !com.Valid() {
+			nb.errChan <- errors.New("received an invalid PoST commitment")
+			return
+		}
+
+		log.Info("finished PoST initialization, commitment: %x",
+			*com)
+
+		nb.nipst.postCommitment = com
+		nb.nipst.persist()
+	}
+
+	for {
+		if nb.stopped() {
+			return
+		}
+
+		// Phase 1
+		if nb.nipst.poetRound == nil {
+			var poetChallenge common.Hash
+
+			// If it's the first NIPST in the chain, use the PoST commitment as
+			// the PoET challenge. Otherwise, use the previous NIPST/ATX hash.
+			if nb.nipst.prev == nil {
+				// TODO(moshababo): check what exactly need to be hashed.
+				poetChallenge = crypto.Keccak256Hash(nb.id, *nb.nipst.postCommitment)
+			} else {
+				// TODO(moshababo): check what exactly need to be hashed.
+				poetChallenge = crypto.Keccak256Hash(*nb.nipst.prev.postProof)
+			}
+
+			log.Info("submitting challenge to PoET proving service, "+
+				"service id: %v, challenge: %x",
+				nb.poetProver.id(), poetChallenge)
+
+			round, err := nb.poetProver.submitChallenge(poetChallenge, nb.duration)
+			if err != nil {
+				nb.errChan <- fmt.Errorf("failed to submit challenge to poet service: %v", err)
+				return
```

Comment:
We obviously need a different retry mechanisms for different phases, and might introduce also manual actions, but that’s out of scope for this function. It will need to be handled by `NIPSTBuilder` client. 
Now i’m focused mainly on the happy flow, but i’ll add the error handling once we’ll get the persistency and p2p integration working.